src/Storages/MergeTree/MergeTreeIndexVectorSimilarity.cpp

```diff
@@ -0,0 +1,492 @@
+    {"L2Distance", unum::usearch::metric_kind_t::l2sq_k},
+    {"cosineDistance", unum::usearch::metric_kind_t::cos_k}};
+
+/// Maps from user-facing name to internal name
+std::unordered_map<String, unum::usearch::scalar_kind_t> quantizationToScalarKind = {
+    {"f32", unum::usearch::scalar_kind_t::f32_k},
+    {"f16", unum::usearch::scalar_kind_t::f16_k},
+    {"i8", unum::usearch::scalar_kind_t::i8_k}};
+
+template<typename T>
+concept is_set = std::same_as<T, std::set<typename T::key_type, typename T::key_compare, typename T::allocator_type>>;
+
+template<typename T>
+concept is_unordered_map = std::same_as<T, std::unordered_map<typename T::key_type, typename T::mapped_type, typename T::hasher, typename T::key_equal, typename T::allocator_type>>;
+
+template <typename T>
+String joinByComma(const T & t)
+{
+    if constexpr (is_set<T>)
+    {
+        return fmt::format("{}", fmt::join(t, ", "));
+    }
+    else if constexpr (is_unordered_map<T>)
+    {
+        String joined_keys;
+        for (const auto & [k, _] : t)
+        {
+            if (!joined_keys.empty())
+                joined_keys += ", ";
+            joined_keys += k;
+        }
+        return joined_keys;
+    }
+    /// TODO once our libcxx is recent enough, replace above by
+    ///      return fmt::format("{}", fmt::join(std::views::keys(t)), ", "));
+    std::unreachable();
+}
+
+}
+
+USearchIndexWithSerialization::USearchIndexWithSerialization(
+    size_t dimensions,
+    unum::usearch::metric_kind_t metric_kind,
+    unum::usearch::scalar_kind_t scalar_kind,
+    UsearchHnswParams usearch_hnsw_params)
+    : Base(Base::make(unum::usearch::metric_punned_t(dimensions, metric_kind, scalar_kind),
+                      unum::usearch::index_dense_config_t(usearch_hnsw_params.m, usearch_hnsw_params.ef_construction, usearch_hnsw_params.ef_search)))
+{
+}
+
+void USearchIndexWithSerialization::serialize(WriteBuffer & ostr) const
+{
+    auto callback = [&ostr](void * from, size_t n)
+    {
+        ostr.write(reinterpret_cast<const char *>(from), n);
+        return true;
+    };
+
+    auto result = Base::save_to_stream(callback);
+    if (result.error)
+        throw Exception::createRuntime(ErrorCodes::INCORRECT_DATA, "Could not save vector similarity index, error: " + String(result.error.release()));
+}
+
+void USearchIndexWithSerialization::deserialize(ReadBuffer & istr)
+{
+    auto callback = [&istr](void * from, size_t n)
+    {
+        istr.readStrict(reinterpret_cast<char *>(from), n);
+        return true;
+    };
+
+    auto result = Base::load_from_stream(callback);
+    if (result.error)
+        /// See the comment in MergeTreeIndexGranuleVectorSimilarity::deserializeBinary why we throw here
+        throw Exception::createRuntime(ErrorCodes::INCORRECT_DATA, "Could not load vector similarity index, error: " + String(result.error.release()) + " Please drop the index and create it again.");
```

Comment:
```suggestion
        throw Exception::createRuntime(ErrorCodes::INCORRECT_DATA, "Could not load vector similarity index, error: {} Please drop the index and create it again."", String(result.error.release()));
```

For the other exceptions too.